### PR TITLE
[fix](Nereids): Optimize BFS Memory Usage to Mitigate Exponential Data Growth

### PR DESCRIPTION
**Pull request by keanji-x**
## Proposed changes
The origin pr is #34948 and the temporary solution is #35408 
In our effort to streamline and optimize dependency handling, we implement the following steps:

* Detect Circular Dependencies: Identify any circular references within functional dependencies. If any are found, we remove the specific dependencies responsible for creating these cycles.
* Clean Up Group By Dependencies: Remove all dependencies listed in the 'group by' clauses to simplify and enhance query performance.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run p0

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run fe

**Comment by XieJiann** [comment]
run p1

**Comment by XieJiann** [comment]
run p0

**Comment by XieJiann** [comment]
run p0

**Comment by XieJiann** [comment]
run p0

**Comment by XieJiann** [comment]
run p0

**Comment by XieJiann** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FuncDeps.java [hunk 1]
```diff
@@ -78,6 +82,32 @@ public boolean isEmpty() {
         return items.isEmpty();
     }
 
+    private void dfs(Set<Slot> parent, Set<Set<Slot>> visited, Set<FuncDepsItem> circleItem) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FuncDeps.java:85]: Seems that current circleItem depends on traversing order, and if the useful item is removed by chance, will we can't do the cooresponding opt at that time?

**Review by xzj7019 — APPROVED** [review]

**Pull request merged**